capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,407 @@
+# SPDX-FileCopyrightText: Copyright DB Netz AG and the capellambse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+__all__ = [
+    "Category",
+    "ElementValidation",
+    "get_passed_and_total",
+    "ModelValidation",
+    "register_rule",
+    "Result",
+    "Results",
+    "Rule",
+    "RuleList",
+    "Rules",
+    "store_result",
+    "Type",
+    "Validation",
+    "Validator",
+]
+
+import abc
+import collections.abc as cabc
+import dataclasses
+import enum
+import functools
+import typing as t
+from itertools import chain
+
+from lxml import etree
+
+import capellambse
+import capellambse.model.modeltypes as mt
+from capellambse import helpers
+from capellambse.model import common as c
+
+
+class Category(mt._StringyEnumMixin, enum.Enum):
+    """A category for a rule."""
+
+    REQUIRED = enum.auto()
+    RECOMMENDED = enum.auto()
+    SUGGESTED = enum.auto()
+
+
+@dataclasses.dataclass
+class Result:
+    """A validation rule result."""
+
+    uuid: helpers.UUIDString
+    category: Category
+    value: bool
+    object: c.GenericElement
+
+    def __repr__(self) -> str:
+        """Return the representation of a result."""
+        obj_repr = self.object._short_repr_()
+        return (
+            f"Result(uuid={self.uuid!r}, category={self.category!r}"
+            f"value={self.value!r}, object={obj_repr})"
+        )
+
+
+ElementType = t.TypeVar("ElementType", bound=c.GenericElement)
+
+Validator = cabc.Callable[
+    [ElementType], t.Union[bool, t.Literal["NotApplicable"]]
+]  # | Callable [[Metric], Result]
+Type = t.Union[
+    str, type[c.GenericElement]
+]  # | type[Metric] | type[AggregateMetric]
+"""Type to be used in capellambse.model.MelodyModel.search(type) or metric
+function the value of which will be validated by the rule."""
+
+
+@dataclasses.dataclass(frozen=True)
+class Rule:
+    """Common (generic) class representing any validation rule."""
+
+    id: str
+    name: str
+    types: list[Type]
+    rationale: str
+    category: Category
+    action: str
+    applicable_to: str
+    validator: Validator
+    hyperlink_further_reading: str | None = None
+
+    def __call__(
+        self, obj: c.GenericElement
+    ) -> bool | t.Literal["NotApplicable"]:
+        return self.validator(obj)
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, str):
+            return self.id == other
+        return hash(self) == hash(other)
+
+
+def _rule_attr_getter(val: str, attr: str) -> Rule | RuleList:
+    all_rules = chain.from_iterable(VALIDATION_RULES.values())
+    rules = [rule for rule in all_rules if getattr(rule, attr) == val]
+    if len(rules) == 1:
+        return rules[0]
+    return RuleList(rules)
+
+
+class RuleList(list[Rule]):
+    """Stores all discovered validation rules."""
```

Comment:
Rules are not discovered, but registered and then queried.  """Stores validation rules.""" is what this  data class  does. Shouldn´t it be decorated as @ataclass even if it does not have any properties to make clear it has no biz logic in it?